rlu/src/stm/mod.rs

```diff
@@ -0,0 +1,659 @@
+            return Ok(self.write.get(tvar).unwrap().clone());
+        }
+
+        if tvar.lock.is_locked() {
+            match tvar.is_locked_by(self.id) {
+                ThreadLockState::Same | ThreadLockState::None => {
+                    tvar.lock.unlock()?;
+                }
+                ThreadLockState::Foreign => {
+                    info!(
+                        "LOAD({:?}): PRECHECK TRANSACTION LOCKED BY ({:?})",
+                        self.id,
+                        tvar.locked_by()
+                    );
+                    return Err(TxError::TransactionLocked);
+                }
+            }
+        }
+
+        let pre_version = tvar.lock.version();
+
+        let data = tvar.original.value.lock().map_err(|e| TxError::LockPresent)?;
+        info!("LOAD({:?}): TVAR CONTENTS ({:?})", self.id, data);
+        let post_version = tvar.lock.version();
+
+        let is_locked = tvar.lock.is_locked();
+        let version_mismatch = pre_version != post_version;
+        let stale_object = pre_version > self.version;
+
+        match is_locked || version_mismatch || stale_object {
+            true => {
+                info!(
+                    "LOAD({:?}): VERSION MISMATCH ({}), STALE OBJECT ({}), PRE_VERSION ({}), TRANSACTION_VERSION ({})",
+                    self.id, version_mismatch, stale_object, pre_version, self.version
+                );
+                Err(TxError::VersionMismatch)
+            }
+            false => Ok((*data).clone()),
+        }
+    }
+
+    /// this writes the value into the transactional log
+    pub fn store(&mut self, tvar: &TVar<T>, value: T) -> Result<(), TxError> {
+        self.write.insert(tvar.clone(), value);
+
+        Ok(())
+    }
+
+    #[inline(always)]
+    fn lock_write_set(&self) -> Result<(), TxError> {
+        let fail_lock_write = AtomicBool::new(false);
+
+        for tvar in self.write.keys() {
+            info!("TRANSACTION({:?}): WRITE LOCK", self.id);
+
+            if tvar.lock.try_lock().is_err() {
+                info!("LOCK WRITE SET({:?}): TRANSACTION LOCKED.", self.id);
+                fail_lock_write.store(true, std::sync::atomic::Ordering::SeqCst);
+                break;
+            }
+
+            match tvar.is_locked_by(self.id) {
+                ThreadLockState::Foreign => {
+                    fail_lock_write.store(true, std::sync::atomic::Ordering::SeqCst);
+                    break;
+                }
+                _ => {
+                    info!(
+                        "TRANSACTION({:?}): LOCK HELD BY US OR NONE ({:?})",
+                        self.id,
+                        tvar.is_locked_by(self.id)
+                    );
+                }
+            }
+
+            // lock to thread id
+            tvar.lock(self.id)?;
+        }
+
+        if fail_lock_write.load(std::sync::atomic::Ordering::SeqCst) {
+            for tvar in self.write.keys() {
+                tvar.lock.unlock()?;
+            }
```

Comment:
Some tvar of the write set may be locked by another thread (since we failed to get all the write locks).
Shouldn't we only release the tvar we managed to get a lock for?

Reply:
hmm the `TVar` here would be a pointer a one managed transactable memory (eg. the HashSet from the test ). Each `Transaction` intending to write the changes to memory needs to lock all `TVars` (eg. references to this one single transactable memory ) in order to be consistent. Since a `TVar` itself contains only atomic references to the actual version and memory, locking one `TVar` locks them all, but by a specific Transaction.